linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java

```diff
@@ -38,65 +41,80 @@
+  // Finding latest version requires metadata from remote repository
+  private final RepositorySystem repositorySystem = RepositoryUtility.newRepositorySystem();
+
+  public static void main(String[] arguments)
+      throws RepositoryException, IOException, LinkageMonitorException {
     if (arguments.length < 1) {
       System.err.println(
           "Please specify BOM coordinates. Example: com.google.cloud:libraries-bom:1.2.1");
       System.exit(1);
     }
     String bomCoordinates = arguments[0];
-    Bom baseline = RepositoryUtility.readBom(bomCoordinates);
-    Bom snapshot = copyWithSnapshot(baseline);
+    new LinkageMonitor().run(bomCoordinates);
+  }
 
-    ImmutableSet<SymbolProblem> problemInBaseline =
+  private void run(String bomCoordinates)
+      throws RepositoryException, IOException, LinkageMonitorException {
+    Bom baseline = RepositoryUtility.readBom(bomCoordinates);
+    ImmutableSet<SymbolProblem> problemsInBaseline =
         LinkageChecker.create(baseline).findSymbolProblems().keySet();
+
+    Bom snapshot = copyWithSnapshot(repositorySystem, baseline);
+
     ImmutableSet<SymbolProblem> problemsInSnapshot =
         LinkageChecker.create(snapshot).findSymbolProblems().keySet();
 
-    if (problemInBaseline.containsAll(problemsInSnapshot)) {
-      // No new symbol problems introduced by snapshot BOM. Returning success.
-      return;
-    } else {
+    Set<SymbolProblem> newErrors = Sets.difference(problemsInSnapshot, problemsInBaseline);
+    if (!newErrors.isEmpty()) {
       // TODO(#683): Display new linkage errors caused by snapshot versions if any
       System.err.println("There are one or more new new linkage errors in snapshot versions:");
-      System.err.println(Sets.difference(problemInBaseline, problemsInSnapshot));
-      System.exit(1);
+      System.err.println(newErrors);
+      int errorSize = newErrors.size();
+      throw new LinkageMonitorException(
+          String.format("Found %d new linkage error%s", errorSize, errorSize > 1 ? "s" : ""));
     }
+    // No new symbol problems introduced by snapshot BOM. Returning success.
   }
 
   /**
    * Returns a copy of {@code bom} replacing its managed dependencies that have locally-installed
    * snapshot versions.
    */
-  private static Bom copyWithSnapshot(Bom bom) throws VersionRangeResolutionException {
+  @VisibleForTesting
+  static Bom copyWithSnapshot(RepositorySystem repositorySystem, Bom bom)
+      throws VersionRangeResolutionException {
     ImmutableList.Builder<Artifact> managedDependencies = ImmutableList.builder();
-
-    RepositorySystem repositorySystem = RepositoryUtility.newFileRepositorySystem();
     RepositorySystemSession session = RepositoryUtility.newSession(repositorySystem);
 
     for (Artifact managedDependency : bom.getManagedDependencies()) {
-
       String snapshotVersion = findSnapshotVersion(repositorySystem, session, managedDependency);
       if (snapshotVersion == null) {
```

Comment:
```java
if (snapshotVersion != null) {
  managedDependency.setVersion(snapshotVersion);
}
managedDependencies.add(managedDependency);
```

Although if `findSnapshotVersion` returned an `Optional` instead of a nullable value, you could even do:

```java
findSnapshotVersion(repositorySystem, session, managedDependency).ifPresent(managedDependency::setVersion);
managedDependencies.add(managedDependency);
```

Reply:
Updated with optional. But `managedDependency::setVersion` returns a new instance with the version.